### PR TITLE
Assignment2

### PR DESCRIPTION
**Pull request by 1bbartels**
Assignment 2 by 1bbartels and EvanBartels

### CODE REVIEW

File: Logger/BaseLoggerMixins.cs [hunk 1]
```diff
@@ -2,6 +2,44 @@
 {
     public static class BaseLoggerMixins
     {
-        
+        public static void Error(this BaseLogger baseLogger, string message, params object[] list)
+        {
+            if (baseLogger is null)
+            {
+                throw new System.ArgumentNullException("null BaseLogger in Error");
```
Comment (randypipkins) [inline Logger/BaseLoggerMixins.cs:9]: Code looks good. Only suggestion I would recommend is to use a nameof() expression rather than a string statement to have a slightly easier time identifying a potential error. Same changes could be made on lines 19, 29, and 39.

**Review by crawfordkid2 — COMMENTED** [review]
There is an existing BaseLogger class. It needs an auto property to hold the class name. This property should be set in the LogFactory using an object initializer. ✔
Create a FileLogger that derives from BaseLogger. It should take in a path to a file to write the log message to. When its Log method is called, it should append messages on their own line in the file. The output should include all of the following:
The current date/time ✔
The name of the class that created the logger ✔
The log level ✔
The message ✔
The format may vary, but an example might look like this "10/7/2019 12:38:59 AM FileLoggerTests Warning: Test message"
The LogFactory should be updated with a new method ConfigureFileLogger. This should take in a file path and store it in a private member. It should use this when instantiating a new FileLogger in its CreateLogger method. ✔
If the file logger has not be configured in the LogFactory, its CreateLogger method should return null. ✔
Inside of BaseLoggerMixins implement extension methods on BaseLogger for
Error, ✔
Warning, ✔
Information, and ✔
Debug. ✔ Each of these methods should take in a string for the message, as well as a parameter array of arguments for the message. Each of these extension methods is expected to be a shortcut for calling the BaseLogger.Log method, by automatically supplying the appropriate LogLevel. These methods should throw an exception if the BaseLogger parameter is null. There are a couple example unit tests to get you started.
Use the nameof() operator when identifying the class name to the logger ✔
Ensure you turn on Warnings as Errors (TreatWarningsAsErrors) ✔
Ensure that you turn on code analysis (EnableNETAnalyzers) ✔
Ensure that you turn on CodeAnalysisTreatWarningsAsErrors ❌
Set LangVersion to 11.0 ✔
Turn on Nullability (Nullable) ✔
All of the above should be unit tested. 
      Notes: 
      - some tests only have an act function, while this will have the test pass, an Assert compares and makes sure the result is the expected result
      - Enable CodeAnalysisTreatWarningsAsErrors, you enabled TreatWarningsAsErrors, there are two specifications in the rubric for the warnings in the enable
      - double check nameOf() usage

File: Logger.Tests/LogFactoryTests.cs [hunk 2]
```diff
@@ -6,6 +6,16 @@ namespace Logger.Tests
...
+        public void CreateLogger_ClassNameIsNull_RetrurnsNull()
+        {
+            // Arrange
+            LogFactory logFactory = new LogFactory();
+
+            // Act
+            logFactory.CreateLogger(null!);
+
+            // Assert
+        }
```
Comment (VeraKorchemnaya) [inline Logger.Tests/LogFactoryTests.cs:19]: You should add a new test in here to make sure that passing in a file path does not return null. In this test you should use the nameof operator. Here is an example of what me and my partner did: 
```suggestion
        }
        [TestMethod]
    [DataRow(nameof(LogFactoryTests))]
    public void CreateLogger_Configured_ReturnsNotNull(string className)
    {
        // Arrange
        LogFactory factory = new();
        factory.ConfigureFileLogger(Path.GetTempFileName());

        // Act
        BaseLogger logger = factory.CreateLogger(className);

        // Assert
        Assert.IsNotNull(logger);
    }
        
        
        
```

**Review by VeraKorchemnaya — COMMENTED** [review]
There is an existing BaseLogger class. It needs an auto property to hold a class name. This property should be set in the LogFactory using an object initializer. ❌ (See comment)
Create a FileLogger that derives from BaseLogger. It should take in a path to a file to write the log message to. When its Log method is called, it should append messages on their own line in the file. The output should include all of the following:
The current date/time ✔
The name of the class that created the logger ❌
The log level ✔
The message ✔
The format may vary, but an example might look like this "10/7/2019 12:38:59 AM FileLoggerTests Warning: Test message"
The LogFactory should be updated with a new method ConfigureFileLogger. This should take in a file path and store it in a private member. It should use this when instantiating a new FileLogger in its CreateLogger method. ✔
If the file logger has not be configured in the LogFactory, its CreateLogger method should return null. ✔
Inside of BaseLoggerMixins implement extension methods on BaseLogger for
Error, ✔
Warning, ✔
Information, and ✔
Debug. ✔ Each of these methods should take in a string for the message, as well as a parameter array of arguments for the message. Each of these extension methods is expected to be a shortcut for calling the BaseLogger.Log method, by automatically supplying the appropriate LogLevel. These methods should throw an exception if the BaseLogger parameter is null. There are a couple example unit tests to get you started.
Use the nameof() operator when identifying the class name to the logger ❌ (This should be done in the testing methods)
Ensure you turn on Warnings as Errors (TreatWarningsAsErrors) ✔
Ensure that you turn on code analysis (EnableNETAnalyzers) ✔
Ensure that you turn on CodeAnalysisTreatWarningsAsErrors ❌
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
Turn on Nullability (Nullable) ✔
All of the above should be unit tested.

File: Logger/FileLogger.cs [hunk 3]
```diff
@@ -0,0 +1,30 @@
+﻿using System;
+using System.IO;
+
+namespace Logger
+{
+    public class FileLogger : BaseLogger
+    {
+        public string FilePath;
+        public FileLogger(string filePath, string name)
```
[the remark this answers is not in the document]
Comment (VeraKorchemnaya) [reply Logger/FileLogger.cs:9]: I was wrong about this, you do want to take in the name like he showed in class.

File: Logger/BaseLoggerMixins.cs (hunk 1 above)
Comment (1bbartels) [reply Logger/BaseLoggerMixins.cs]: The argument has been updated to use nameof()

File: Logger/BaseLoggerMixins.cs [hunk 4]
```diff
@@ -2,6 +2,44 @@
 {
     public static class BaseLoggerMixins
     {
-        
+        public static void Error(this BaseLogger baseLogger, string message, params object[] list)
```
[the remark this answers is not in the document]
Comment (1bbartels) [reply Logger/BaseLoggerMixins.cs]: I like the rename suggestion, so I'll use it.

File: Logger/LogFactory.cs [hunk 5]
```diff
@@ -2,10 +2,22 @@
...
         {
-            
-            return null;
+            if(FilePath is null)
+            {
+                return null!;
+            }
+
+            FileLogger output = new FileLogger(FilePath) {ClassName = className};
+            return output;
```
Comment (mmwoodfo) [inline Logger/LogFactory.cs:15]:
```suggestion
            return new FileLogger(FilePath) {ClassName = className};
```
I'd recommend simplifying your code to the above.

File: Logger/FileLogger.cs [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+            if(filePath is null)
+            {
+                throw new System.ArgumentNullException(nameof(filePath));
+            }
+            _filePath = filePath;    
+        }
+
+        public override void Log(LogLevel logLevel, string message)
+        {
+            using StreamWriter sw = File.AppendText(message);
```
Comment (mmwoodfo) [inline Logger/FileLogger.cs:20]: Wrapping this using statement around your `StreamWriter` will close it for you automatically once it goes out of scope. So you don't need to call `Close` on line 23 as long as you're using the `using` keyword.

File: Logger/LogFactory.cs [hunk 7]
```diff
@@ -2,10 +2,22 @@
...
     {
+        private string? FilePath;
+
         public BaseLogger CreateLogger(string className)
         {
-            
-            return null;
+            if(FilePath is null)
+            {
+                return null!;
```
Comment (mmwoodfo) [inline Logger/LogFactory.cs:11]: Make your method signature nullable, don't use `null!`

File: Logger/LogFactory.cs [hunk 8]
```diff
@@ -2,10 +2,22 @@
 {
     public class LogFactory
     {
+        private string? FilePath;
+
         public BaseLogger CreateLogger(string className)
```
Comment (mmwoodfo) [inline Logger/LogFactory.cs:7]: Make this nullable
```suggestion
        public BaseLogger? CreateLogger(string className)
```

File: Logger/BaseLoggerMixins.cs [hunk 9]
```diff
@@ -2,6 +2,42 @@
 {
     public static class BaseLoggerMixins
     {
-        
+        public static void Error(this BaseLogger baseLogger, string message, params object[] args)
+        {
+            if (baseLogger is null)
```
Comment (mmwoodfo) [inline Logger/BaseLoggerMixins.cs:7]: Good job, something to consider, you could pull this if check out into a private static helper method that takes a BaseLogger and does this check for you, rather than re-writing it 3 more times.
